src/quantum_gates/_simulation/backend.py

```diff
@@ -212,6 +215,203 @@ def _chunk_list(self, l: list, min_chunk_size: int, optimal_chunk_size: int) ->
+             qubit_layout(list): Layout of the qubit used in the optimizator
+
+        Returns:
+            The propagated statevector.
+        """
+        assert len(mp_list) > 0, f"Expected non empty matrix product list, but found {mp_list}."
+        psi1 = copy.deepcopy(psi0)
+ 
+        mp_list_opt = Optimizator(level_opt= level_opt, circ_list= mp_list, qubit_list=qubit_layout).optimize()
+
+        for item in mp_list_opt:
+
+            q_list = list(range(self.nqubit)) # list of indexes of all qubit in the circuit
+            q_used = item[1] # list of indexes of the qubit used in this moment
+
+            if len(item[1]) == 1: # check if the current is a 1 qubit
+                q = item[1][0]
+                q_list.remove(q) # remove index of the qubit from the list
+            elif len(item[1]) == 2: # check if the current is a 2 qubit gates
+                q1 = item[1][0]
+                q2 = item[1][1]
+                q_list.remove(q1) # remove index of the qubit from the list
+                q_list.remove(q2)
+
+            q_n_used = q_list
+
+            k = len(q_n_used)
+
+            if k == 0: # in case of 1 or 2 qubits circuit
+                U = self.create_dense(item =item, q_used = q_used, q_n_used = q_n_used)
+                psi1 = U @ psi1
+
+            elif k > 0:
+                U = self.create_sparse(item = item, q_n_used = q_n_used, q_used = q_used, N = self.nqubit)
+                psi1 = U.dot(psi1)
+
+        return psi1
+    
+    def create_sparse(self, item : list, q_n_used : list , q_used : list, N : int):
+        """Given a gate applied on some qubits the function return a sparse array to use in the statevector function
+
+        Args:
+            item (list): List representing the entries of the gate and in whithc qubit is applied
+            q_n_used (list): list of not used qubit
+            q_used (list): list of used qubit
+            N (int): Number of total qubit
+
+        Raises:
+            ValueError: If the number of total qubit doesn't match the sum of used and not used qubit there is a problem
+
+        Returns:
+            Sparse matrix: Sparse matrix representing the gate
+        """
+
+        k = len(q_n_used) # number of not used qubit
+        m = len(q_used) # m = n-k number of used qubit
+
+        if k+m != N:
+            raise ValueError(f"You lose some qubit along the way, total number of qubit is {N} and the sum of used and not used qubit is {k+m}")
+
+        # quantities for the sparse matrix
+        data = []
+        row_indices = []
+        col_indices = []
+
+        for i in range(2**k): # generate only the non zero element in the matrix according to the number of not used qubit
+                k_str = f"{i*(2**k+1):0{2*k}b}"
+                for j in range(2**(2*(m))):
+                    m_str = f"{j:0{2*(m)}b}"
+                    n_str = self.join_str(k_str, m_str, q_n_used, q_used ,k ,m) # merge string
```

Comment:
@paolodr98 What is the idea behind the join_str method?

Reply:
The idea is to calculate only the binary strings for the non-zero elements of the matrix: First, all strings representing the unused qubits are generated and stored in k_str. Then, the strings for the used qubits are generated and stored in m_str. In the `join_str` method, the bit of these strings are placed in the correct position, according to which qubits are used and which are not. I'll provide you an example later during the meeting